### PR TITLE
Add `StreamProvider` for configuring `StreamTable`

### LINKED ISSUE #10599: Make the configuration for `StreamTable` more generic to support more stream sources
### Is your feature request related to a problem or challenge?

I am working on a websocket `TableProvider` and initially I went about creating my own `TableProvider` but then after reviewing the `StreamTable` and `StreamConfig` it seems it was made with the potential to be generic at some point.

Ref: "Currently only reading from / appending to a single file in-place is supported, but other stream sources and sinks may be added in future."

So I decided to try making a more generic interface for this.  As a first step I created the new interface and ported the existing implementation to use this (I will have a PR up shortly).  I havent finished cleaning up yet but i think the API is in good enough shape to start getting feedback.

The TLDR of the new interface is that there is a new trait `StreamSource` which provides the `Schema` and `RecordBatchReader` and `RecordBatchWriter`.  Then there are implementations like `FileStreamSource` or `WebsocketStreamSource`.

Assuming this is going in the right direction I can cleanup more, get this merged, and then follow on with an implementation of `WebsocketStreamSource`.

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by matthewmturner**
## Which issue does this PR close?

Closes #10599 

## Are these changes tested?

Not yet, but I will add tests if there is agreement the API is going in the right direction

## Are there any user-facing changes?

Yes, new interface for `StreamTable`

### CODE REVIEW

**Comment by matthewmturner** [comment]
@metesynnada @mustafasrepo i believe you were both involved in the `StreamTable` implementation so im interested in getting your views if this is going in the right direction towards a more generic `StreamTable`. (theres plenty of cleanup to do still but the general API is there)

File: datafusion/core/src/datasource/stream.rs [hunk 1]
```diff
@@ -103,19 +105,29 @@ impl FromStr for StreamEncoding {
     }
 }
 
-/// The configuration for a [`StreamTable`]
+pub trait StreamSource: std::fmt::Debug + Send + Sync {
```
Comment (alamb) [inline datafusion/core/src/datasource/stream.rs:108]: Perhaps some documentation about what this trait is meant to represent would help

**Review by alamb — COMMENTED** [review]
Thanks @matthewmturner  -- this looks pretty neat.

I wonder if there is some example we could add to `datafusion-examples` that would show this new API in action? That might help motivate what the API changes are for / help this PR review along as well as help others see the feature

**Comment by matthewmturner** [comment]
@alamb i would be happy to add example - for this PR it would likely just be copying from https://github.com/apache/datafusion/blob/main/datafusion/core/tests/fifo.rs.  I will get to that shortly.

Ultimately the motivation here is to start working towards a more generic interface where `StreamTable` can be used for multiple stream types as opposed to just files (such as websockets, kafka, etc. or maybe thats a non-goal, in which case i can just close this).  Ive seen that it can require lots of custom code (formats and providers) to implement streaming tables and im hoping to simplify that.

**Comment by matthewmturner** [comment]
Im hoping to get to a similar API as `ListingTable`.

`ListingTable` => `ListingTableConfig` => `FileFormat`

Where `ListingTable` and `ListingTableConfig` are provided by datafusion and if you want to extend then you can just implement a `FileFormat` and benefit from the `ListingTable` and `ListingTableConfig` machinery.

In the case of streams i had in mind: 

`StreamTable` => `StreamTableConfig` => `StreamProvider` (maybe i rename to `StreamFormat` for consistency?)

**Comment by matthewmturner** [comment]
@alamb i have a working example now. i have idea to update it to show more of the streaming nature (i.e. write to the fifo and get batches multiple times) but wont have time for that today.  Do you have thoughts in general on whether this type of API could be supported?

File: datafusion/core/src/datasource/stream.rs [hunk 2]
```diff
@@ -103,19 +105,46 @@ impl FromStr for StreamEncoding {
...
 }
 
-/// The configuration for a [`StreamTable`]
+/// The StreamProvider trait is used as a generic interface for reading and writing from streaming
+/// data sources (such as FIFO, Websocket, Kafka, etc.).  Implementations of the provider are
+/// responsible for providing a `RecordBatchReader` and optionally a `RecordBatchWriter`.
+pub trait StreamProvider: std::fmt::Debug + Send + Sync {
+    /// Get a reference to the schema for this stream
+    fn schema(&self) -> SchemaRef;
+    /// Needed for `PartitionStream` - maybe there is a better way to do this.
```
Comment (alamb) [inline datafusion/core/src/datasource/stream.rs:114]: I think changing `schema` to return `&SchemaRef` would be enough -- then anyone who needed an owned `SchemaRef` could clone the result. I don't think we need separate methods

File: datafusion-examples/examples/file_stream_provider.rs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    // Specify the ordering:
+    let order = vec![vec![datafusion_expr::col("a1").sort(true, false)]];
+
+    let provider = fifo_table(schema.clone(), fifo_path, order.clone());
+    ctx.register_table("fifo", provider)?;
+
+    let df = ctx.sql("SELECT * FROM fifo").await.unwrap();
+    let mut stream = df.execute_stream().await.unwrap();
+
+    futures::future::join_all(tasks).await;
```
Comment (alamb) [inline datafusion-examples/examples/file_stream_provider.rs:164]: I wonder if you need to join all here (it kind of makes the streaming nature of this example less clear) -- could you potentially do after you read the batches?

File: datafusion-examples/examples/file_stream_provider.rs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    // Create session context
+    let config = SessionConfig::new()
+        .with_batch_size(TEST_BATCH_SIZE)
+        .with_collect_statistics(false)
+        .with_target_partitions(1);
+    let ctx = SessionContext::new_with_config(config);
+    let tmp_dir = TempDir::new()?;
+    let fifo_path = create_fifo_file(&tmp_dir, "fifo_unbounded.csv")?;
+
+    let mut tasks: Vec<JoinHandle<()>> = vec![];
```
Comment (alamb) [inline datafusion-examples/examples/file_stream_provider.rs:133]: BTW using a `JoinSet` https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html might be easier here (it encapsulates managing tasks like this)

**Review by alamb — COMMENTED** [review]
I think this API looks pretty good @matthewmturner  thank you

@metesynnada  or @ozankabak or @devinjdangelo  do you have any thoughts / suggestions on these patterns?

**Comment by ozankabak** [comment]
@berkaysynnada PTAL

**Comment by berkaysynnada** [comment]
I will review it in detail tomorrow

File: datafusion/core/src/datasource/stream.rs [hunk 5]
```diff
@@ -103,19 +105,46 @@ impl FromStr for StreamEncoding {
     }
 }
 
-/// The configuration for a [`StreamTable`]
+/// The StreamProvider trait is used as a generic interface for reading and writing from streaming
+/// data sources (such as FIFO, Websocket, Kafka, etc.).  Implementations of the provider are
+/// responsible for providing a `RecordBatchReader` and optionally a `RecordBatchWriter`.
+pub trait StreamProvider: std::fmt::Debug + Send + Sync {
```
Comment (matthewmturner) [inline datafusion/core/src/datasource/stream.rs:111]: `StreamFormat` might be a better name here, to align with `FileFormat`

File: datafusion-examples/examples/file_stream_provider.rs (hunk 4 above)
Comment (matthewmturner) [reply datafusion-examples/examples/file_stream_provider.rs:133]: Yes, I was actually planning on making this change. Thanks for reminding me

File: datafusion-examples/examples/file_stream_provider.rs (hunk 3 above)
Comment (matthewmturner) [reply datafusion-examples/examples/file_stream_provider.rs:164]: Will try that out

File: datafusion/core/src/datasource/stream.rs (hunk 5 above)
Comment (berkaysynnada) [reply datafusion/core/src/datasource/stream.rs:111]: I am not sure people associate location with 'Format'. FileFormat does not deal with location. I think StreamProvider is descriptive enough.

**Review by berkaysynnada — COMMENTED** [review]
After addressing @alamb's feedback, this PR looks good and separates responsibilities with an additional layer. However, `FileStreamProvider` currently handles only the CSV format. I suggest focusing more on file format abstraction. Thanks for the example; it makes understanding and usage easier.

File: datafusion/core/src/datasource/stream.rs (hunk 5 above)
Comment (matthewmturner) [reply datafusion/core/src/datasource/stream.rs:111]: i viewed location as specific to the `FileStreamProvider` and `Format` as a concept that provided reading and writing capabilities, similar to `FileFormat`.  But i dont feel strongly either way so im fine to leave at `StreamProvider`.

**Review by alamb — APPROVED** [review]
This PR looks good to me. Are we waiting on anything else? Otherwise I think it is good to go from my perspective. 

Thank you @matthewmturner and @berkaysynnada  for your code and review

**Review by ozankabak — APPROVED** [review]
LGTM as well

**Comment by matthewmturner** [comment]
thank you @alamb @berkaysynnada @ozankabak

**Comment by alamb** [comment]
Thanks again everyone@

**Pull request merged**